File: Digito Verificador/Modulo 10/DV Gerador/gerardig_mod10.c
// GerarMatricula
#include <stdio.h>
#include <string.h>
#include <ctype.h>

int validarMatricula(const char* MATRICULA)
{
  if (strlen(MATRICULA) != 7)
    return 0;
  if (!isdigit(MATRICULA[0]) || !isdigit(MATRICULA[6]))
    return 0;
  if (MATRICULA[3] != '.')
    return 0;
  for (int I = 1; I <= 2; I++)
    if (!isdigit(MATRICULA[I]) || !isdigit(MATRICULA[I + 4]))
      return 0;
  return 1;
}

void removerCaractere(char CADEIA[], char CARACTERE)
{
  char *POS1 = CADEIA;
  char *POS2 = CADEIA;
  while (*POS1)
  {
    *POS2 = *POS1++;
    POS2 += (*POS2 != CARACTERE);
  }
  *POS2 = '\0';
}

int mod10(const char* MATRICULA)
{
  int MATRIC_INT[6];
  int SOMA = 0;
  for (int I = 0; I <= 5; I++)
    MATRIC_INT[I] = MATRICULA[I] - 48;  // ASC 48 = "0"
  for (int I = 0; I <= 5; I++)
  {
    if (I % 2 != 0 && MATRIC_INT[I] * 2 > 9)
      SOMA += MATRIC_INT[I] * 2 - 9;
    else if (I % 2 != 0 && MATRIC_INT[I] * 2 <= 9)
      SOMA += MATRIC_INT[I] * 2;
    if (I % 2 == 0)
      SOMA += MATRIC_INT[I];
  }
  int DV = (SOMA / 10 + 1) * 10 - SOMA;
  return DV;
}

int main(void)
{
  
  char MATRIC_CPT[8];
  char MATRIC_SPT[8];
  int DV;

  do
  {
    printf("Informe matricula no formato 999.999: ");
    scanf("%s", MATRIC_CPT);
    getchar();
    if (!validarMatricula(MATRIC_CPT))
    {
      printf("\n");
      printf("Matricula invalida. Tente novamente.\n");
      printf("\n\n");
    }
  } while (!validarMatricula(MATRIC_CPT));
  
  strcpy(MATRIC_SPT, MATRIC_CPT);
  removerCaractere(MATRIC_SPT, '.');
  DV = mod10(MATRIC_SPT);
  printf("\n");
  printf("Matricula com DV = %s-%d\n", MATRIC_CPT, DV);
  printf("\n");
  printf("Tecle <Enter> para encerrar... ");
  getchar();
  return 0;
}

File: Digito Verificador/Modulo 10/DV Gerador/gerardig_mod10.go
// GerarMatricula
package main

import (
	"fmt"
	"strconv"
	"strings"
	"unicode"
)

func validarMatricula(MATRICULA string) bool {
	if len(MATRICULA) != 7 {
		return false
	}
	if !isdigit(MATRICULA[0]) || !isdigit(MATRICULA[6]) {
		return false
	}
	if MATRICULA[3] != '.' {
		return false
	}
	for I := 1; I <= 2; I++ {
		if !isdigit(MATRICULA[I]) || !isdigit(MATRICULA[I + 4]) {
			return false
		}
	}
	return true
}

func removerCaractere(CADEIA string, CARACTERE string) string {
	return strings.Replace(CADEIA, CARACTERE, "", -1)
}

func mod10(MATRICULA string) int {
	SOMA := 0
	for I := 0; I <= 5; I++ {
		DIGITO, _ := strconv.Atoi(string(MATRICULA[I]))
		if I % 2 != 0 && DIGITO * 2 > 9 {
			SOMA += DIGITO*2 - 9
		} else if I % 2 != 0 && DIGITO * 2 <= 9 {
			SOMA += DIGITO * 2
		} else {
			SOMA += DIGITO
		}
	}
	DV := (SOMA / 10 + 1) * 10 - SOMA
	return DV
}

func isdigit(CARACTERE uint8) bool {
	return unicode.IsDigit(rune(CARACTERE))
}

func main() {
	var MATRIC_CPT string
	var MATRIC_SPT string
	var MATRIC_INT [6]int
	var I, DV int

	for {
		fmt.Print("Informe matricula no formato 999.999: ")
		fmt.Scanln(&MATRIC_CPT)
		if !validarMatricula(MATRIC_CPT) {
			fmt.Println()
			fmt.Println("Matricula invalida. Tente novamente.\n")
			fmt.Println()
			continue
		}
		break
	}
	
	MATRIC_SPT = MATRIC_CPT
	MATRIC_SPT = removerCaractere(MATRIC_SPT, ".")
	for I = 0; I <= 5; I++ {
		MATRIC_INT[I], _ = strconv.Atoi(string(MATRIC_SPT[I]))
	}
	DV = mod10(MATRIC_SPT)
	fmt.Println()
	fmt.Printf("Matricula com DV = %s-%d\n", MATRIC_CPT, DV)
	fmt.Println()
	fmt.Println("Tecle <Enter> para encerrar...")
	fmt.Scanln()
}
